File: backend/src/modules/transportation/lines/lines.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../../prisma/prisma.service';
import { fail, paginationHeader, successList, successWrite } from '../../../common/response/base-response';
import { NotificationsService } from '../notifications/notifications.service';

@Injectable()
export class LinesService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly notifications: NotificationsService,
  ) {}

  /** GET getAllTransportationLine — paginated lines ({ Id, Name, RouteNum }); optional name search. */
  async getAll(pageNo = 1, noOfItems = 20, name?: string) {
    const where: Record<string, unknown> = { active: true };
    if (name) where.lineName = { contains: name };
    const [total, rows] = await this.prisma.$transaction([
      this.prisma.transportationLine.count({ where }),
      this.prisma.transportationLine.findMany({
        where,
        include: { _count: { select: { routes: { where: { active: true } } } } },
        orderBy: { id: 'desc' },
        skip: (pageNo - 1) * noOfItems,
        take: noOfItems,
      }),
    ]);
    const data = rows.map((l) => ({
      Id: l.id,
      Name: l.lineName,
      RouteNum: l._count.routes, // count of active child routes
      IsApproved: l.isApproved ?? false,
    }));
    return successList(data, paginationHeader(pageNo, noOfItems, total));
  }

  /** POST AddTransportationLine — create a line. */
  async add(lineName: string, userId: number) {
    if (!lineName?.trim()) return fail('Err101', 'LineName is required');
    const created = await this.prisma.transportationLine.create({
      data: { lineName, creationBy: userId, modifiedBy: userId },
    });
    await this.notifications.notifySuperAdmins(userId, {
      title: 'إضافة محطة تجمع جديدة',
      description: 'بحاجة لاعتماد',
      entityType: 'line',
      entityId: created.id,
    });
    return successWrite(created.id);
  }

  /** POST UpdateTransportationLine — rename a line. */
  async update(id: number, lineName: string, userId: number) {
    if (!id) return fail('Err101', 'Id is required');
    if (!lineName?.trim()) return fail('Err101', 'LineName is required');
    await this.prisma.transportationLine.update({ where: { id }, data: { lineName, modifiedBy: userId } });
    return successWrite(id);
  }

  /** POST DeleteTransportationLine — delete (Id header). Blocks with a friendly
   *  message when the gathering station still has routes (FK guard). */
  async remove(id: number) {
    if (!id) return fail('Err101', 'Id is required');
    const routeCount = await this.prisma.transportationVehicleRoute.count({
      where: { transportationLineId: id },
    });
    if (routeCount > 0) {
      return fail(
        'Err102',
        'لا يمكن حذف محطة التجمع لارتباطها بخطوط. احذف الخطوط المرتبطة بها أولاً.',
      );
    }
    await this.prisma.transportationLine.delete({ where: { id } });
    return successWrite(id);
  }

  /** POST ApproveTransportationLine — approve/unapprove (Id + Approve headers). */
  async approve(id: number, approve: boolean, userId: number) {
    if (!id) return fail('Err101', 'Id is required');
    await this.prisma.transportationLine.update({
      where: { id },
      data: { isApproved: approve, approvedBy: userId },
    });
    return successWrite(id);
  }
}

File: backend-dotnet/GarasTransport.Api/Modules/Transportation/Lines/LinesService.cs
using GarasTransport.Api.Common;
using GarasTransport.Api.Modules.Transportation.Notifications;
using Microsoft.EntityFrameworkCore;

namespace GarasTransport.Api.Modules.Transportation.Lines;

public class LinesService
{
    private readonly AppDbContext _db;
    private readonly NotificationsService _notifications;

    public LinesService(AppDbContext db, NotificationsService notifications)
    {
        _db = db;
        _notifications = notifications;
    }

    /// <summary>GET getAllTransportationLine — paginated lines ({ Id, Name, RouteNum }); optional name search.</summary>
    public async Task<object> GetAll(int pageNo = 1, int noOfItems = 20, string? name = null)
    {
        var query = _db.TransportationLines.Where(l => l.Active);
        if (!string.IsNullOrEmpty(name))
            query = query.Where(l => l.LineName.Contains(name));

        var total = await query.CountAsync();
        var rows = await query
            .OrderByDescending(l => l.Id)
            .Skip((pageNo - 1) * noOfItems)
            .Take(noOfItems)
            .Select(l => new
            {
                Id = l.Id,
                Name = l.LineName,
                RouteNum = l.Routes.Count(r => r.Active), // count of active child routes
                IsApproved = l.IsApproved ?? false,
            })
            .ToListAsync();

        return Envelope.SuccessList(rows, Envelope.Pagination(pageNo, noOfItems, total));
    }

    /// <summary>POST AddTransportationLine — create a line.</summary>
    public async Task<object> Add(string? lineName, int userId)
    {
        if (string.IsNullOrWhiteSpace(lineName)) return Envelope.Fail("Err101", "LineName is required");
        var created = new TransportationLine { LineName = lineName, CreationBy = userId, ModifiedBy = userId };
        _db.TransportationLines.Add(created);
        await _db.SaveChangesAsync();
        await _notifications.NotifySuperAdmins(userId, new NotifyOptions
        {
            Title = "إضافة محطة تجمع جديدة",
            Description = "بحاجة لاعتماد",
            EntityType = "line",
            EntityId = created.Id,
        });
        return Envelope.SuccessWrite(created.Id);
    }

    /// <summary>POST UpdateTransportationLine — rename a line.</summary>
    public async Task<object> Update(int id, string? lineName, int userId)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        if (string.IsNullOrWhiteSpace(lineName)) return Envelope.Fail("Err101", "LineName is required");
        var line = await _db.TransportationLines.FirstOrDefaultAsync(l => l.Id == id);
        if (line == null) return Envelope.Fail("Err142", "Line not found");
        line.LineName = lineName;
        line.ModifiedBy = userId;
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }

    /// <summary>POST DeleteTransportationLine — delete (Id header). Blocks when the
    /// gathering station still has routes (FK guard).</summary>
    public async Task<object> Remove(int id)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        var routeCount = await _db.TransportationVehicleRoutes.CountAsync(r => r.TransportationLineId == id);
        if (routeCount > 0)
            return Envelope.Fail("Err102", "لا يمكن حذف محطة التجمع لارتباطها بخطوط. احذف الخطوط المرتبطة بها أولاً.");
        var line = await _db.TransportationLines.FirstOrDefaultAsync(l => l.Id == id);
        if (line == null) return Envelope.Fail("Err142", "Line not found");
        _db.TransportationLines.Remove(line);
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }

    /// <summary>POST ApproveTransportationLine — approve/unapprove (Id + Approve headers).</summary>
    public async Task<object> Approve(int id, bool approve, int userId)
    {
        if (id == 0) return Envelope.Fail("Err101", "Id is required");
        var line = await _db.TransportationLines.FirstOrDefaultAsync(l => l.Id == id);
        if (line == null) return Envelope.Fail("Err142", "Line not found");
        line.IsApproved = approve;
        line.ApprovedBy = userId;
        await _db.SaveChangesAsync();
        return Envelope.SuccessWrite(id);
    }
}
